### PR TITLE
NETOBSERV-1052: Add monitoring dashboard controller to create dashboard configmap

### PR DESCRIPTION
**Pull request by OlivierCazade**
This PR add a new controller to manage a configmap to deploy monitoring dashboard.

The configmap is deployed as long as the controlPlaneTopology is not set to external.

The dashboard itself is there as an example but I wanted to open the PR to start getting review on the controller part.

I am going to look more precisely about what would be the most valuable to display in the dashboard.

### CODE REVIEW

**Comment by candita** [comment]
@OlivierCazade there is a verify error - no linkage to new e2e test.

File: pkg/operator/controller/monitoring-dashboard/dashboard.json
Comment (frobware) [inline pkg/operator/controller/monitoring-dashboard/dashboard.json:1]: I'm wondering what our mid- to longer-term plans are here? Could we encapsulate the metrics in a dashboard-metrics.go (i.e., appropriate types and values) and use `go generate` during the build to generate this JSON file? It would be easier to make larger or more wholesale changes to the output (if necessary).

File: test/e2e/dashboard_test.go
Comment (frobware) [inline test/e2e/dashboard_test.go:15]: This test will fail in CI if the following holds true:

```go
func desiredMonitoringDashboard(ctx context.Context, infraStatus configv1.InfrastructureStatus) *corev1.ConfigMap {
	// If control plane topology is set to external, we do not deploy the dashboard
	if infraStatus.ControlPlaneTopology == configv1.ExternalTopologyMode {
		return nil
	}
```

We need to do the same check and not run the remainder of this test if there is no dashboard deployed.

File: pkg/operator/controller/monitoring-dashboard/monitoring_dashboard.go
Comment (frobware) [reply pkg/operator/controller/monitoring-dashboard/monitoring_dashboard.go:38]: I notice that none of our other controllers make a call to `SetResourceVersion` - asking why we need to do this here?

File: pkg/operator/controller/monitoring-dashboard/dashboard.json
Comment (OlivierCazade) [reply pkg/operator/controller/monitoring-dashboard/dashboard.json:1]: To my knowledge Grafana does not provide tools to generate such dashboard programmatically. Third party tools libraries exist but we did not test them yet.

Please note that this file contain metrics but also many details about the dashboard presentation (Section, subsection, graph width...)

File: pkg/operator/controller/monitoring-dashboard/monitoring_dashboard.go
Comment (OlivierCazade) [reply pkg/operator/controller/monitoring-dashboard/monitoring_dashboard.go]: Thanks, yes, it looks like a naming convention that I missed.

File: test/e2e/all_test.go
Comment (frobware) [inline test/e2e/all_test.go:84]: As `TestDashboardCreation` mutates the global `infracConfig` object we should move this test to the "serial" suite.

File: pkg/operator/controller/monitoring-dashboard/monitoring_dashboard.go
Comment (frobware) [inline pkg/operator/controller/monitoring-dashboard/monitoring_dashboard.go:73]: We reference "dashboard.json" quite a lot--can we capture the name as an exported constant so that the e2e test can also refer to the name symbolically:

```console
% ag dashboard.json
pkg/operator/controller/monitoring-dashboard/monitoring_dashboard.go
70:// dashboardJSON is the string representation of the embedded 'dashboard.json' file.
72://go:embed dashboard.json
91:                     "dashboard.json": dashboardJSON,

pkg/operator/controller/monitoring-dashboard/monitoring_dashboard_test.go
23:                     "dashboard.json": dashboardJSON,
84:                                             "dashboard.json": "corrupted text",
105:                                            "dashboard.json":  dashboardJSON,
125:                                            "dashboard.json": dashboardJSON,
146:                                            "dashboard.json": dashboardJSON,
226:                                            "dashboard.json": dashboardJSON,
241:                                            "dashboard.json": dashboardJSON,

test/e2e/dashboard_test.go
42:             "dashboard.json": "",
52:             dashboard, ok := dashboardCM.Data["dashboard.json"]
65:             "dashboard.json": "",
```

**Review by frobware — CHANGES_REQUESTED** [review]

File: test/e2e/dashboard_test.go
Comment (frobware) [inline test/e2e/dashboard_test.go:57]: I think this could be a little simpler: 

```suggestion
	err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
		if err := kclient.Get(context.TODO(), monitoringdashboard.ConfigMapName(), dashboardCM); err != nil {
			t.Logf("Retrying due to error: %v", err)
			return false, nil
		}
		if dashboard, ok := dashboardCM.Data["dashboard.json"]; ok && dashboard != "" {
			return true, nil
		}
		t.Log("Awaiting dashboard update...")
		return false, nil
	})
```

Comment (OlivierCazade) [reply test/e2e/dashboard_test.go]: Thanks!
Github did not allow me to directly commit your changes because of changes to add the "dashbord.json" constant but I added them to the last commit.

Comment (frobware) [reply test/e2e/dashboard_test.go]: No problem.

File: pkg/operator/controller/monitoring-dashboard/controller.go
Comment (frobware) [inline pkg/operator/controller/monitoring-dashboard/controller.go:24]: This appears to be unused so we should drop it.

**Comment by frobware** [comment]
/lgtm
/hold

The [EP](https://github.com/openshift/enhancements/pull/1431) needs approval.

**Review by frobware — APPROVED** [review]

**Comment by frobware** [comment]
/label tide/merge-method-squash
/lgtm
/approve

Cancelling the hold per: https://redhat-internal.slack.com/archives/C05FWJDFPL4/p1701189670070019?thread_ts=1697538070.044369&cid=C05FWJDFPL4

/hold cancel

**Comment by frobware** [comment]
Note: The `e2e-gcp-operator` job is permanently failing at the moment; see slack thread https://redhat-internal.slack.com/archives/CBZHF4DHC/p1694792730351859.

Bug# pending.

**Comment by frobware** [comment]
> Note: The `e2e-gcp-operator` job is permanently failing at the moment; see slack thread https://redhat-internal.slack.com/archives/CBZHF4DHC/p1694792730351859.
> 
> Bug# pending.

https://issues.redhat.com/browse/OCPBUGS-24044

**Comment by Miciah** [comment]
[e2e-gcp-operator failed again](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/962/pull-ci-openshift-cluster-ingress-operator-master-e2e-gcp-operator/1729570286665207808) because `TestUnmanagedDNSToManagedDNSInternalIngressController` failed, which is a known issue ([OCPBUGS-24044](https://issues.redhat.com/browse/OCPBUGS-24044)), and this PR doesn't change anything that would impact GCP specifically, so we can override the test failure.  
/override ci/prow/e2e-gcp-operator

File: pkg/operator/controller/monitoring-dashboard/dashboard.json
Comment (simonpasquier) [reply pkg/operator/controller/monitoring-dashboard/dashboard.json:1]: There are tools like https://github.com/grafana/grafonnet to generate dashboards programmatically but from my personal experience, nothing really satisfying. Especially if you don't plan to reuse/share the dashboards with other projects, there's little incentive doing so.

**Comment by frobware** [comment]
I added some tooling to verify the server response errors here: https://github.com/frobware/openshift-ingress-dashboard-test.

**Pull request merged**
